### PR TITLE
added logout and redis module

### PR DESCRIPTION
**Pull request by paw9810**

### CODE REVIEW

File: src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -26,11 +28,19 @@ export class AuthService {
         return null
     }
 
-    login(user: User) {
+    async login(user: User) {
         const payload = { userName: user.userName, sub: user.userId }
+        const token = this.jwtService.sign(payload)
+        await this.redisService.setToken(token, user.userId)
```
Comment (WadasGrzegorz) [inline src/modules/auth/auth.service.ts:34]: use key like `access-token-userUUID`

File: src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -26,11 +28,19 @@ export class AuthService {
...
         return {
-            accessToken: this.jwtService.sign(payload),
+            accessToken: token,
         }
     }
+
+    async logout(userId: number) {
+        await this.redisService.delToken(userId)
+
+        return 'Successfully logged out'
```
Comment (WadasGrzegorz) [inline src/modules/auth/auth.service.ts:44]: no magic string - if you need that You can create const with translations

File: src/modules/auth/strategies/jwt.strategy.ts [hunk 3]
```diff
@@ -1,13 +1,17 @@
...
 import { ExtractJwt, Strategy } from 'passport-jwt'
 import { Jwt } from 'lib/enums'
+import { RedisService } from 'modules/redis'
 import { Payload } from './types'
 
 @Injectable()
 export class JwtStrategy extends PassportStrategy(Strategy) {
-    constructor(private configService: ConfigService) {
+    constructor(
+        private configService: ConfigService,
```
Comment (WadasGrzegorz) [inline src/modules/auth/strategies/jwt.strategy.ts:12]: in this place also readonly

File: src/modules/auth/strategies/jwt.strategy.ts [hunk 4]
```diff
@@ -16,6 +20,12 @@ export class JwtStrategy extends PassportStrategy(Strategy) {
     }
 
     async validate(payload: Payload) {
+        const token = await this.redisService.getToken(payload.sub)
```
Comment (WadasGrzegorz) [inline src/modules/auth/strategies/jwt.strategy.ts:23]: create in authService method like `validateAccessToken(userUUID: string, token: string)`

File: src/modules/redis/redis.service.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+export class RedisService {
+    constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}
+
+    async setToken(token: string, userId: number) {
+        const ttl: number = Jwt.TTL
+        const key = `accessToken_${userId}`
+        await this.cacheManager.set(key, token, { ttl })
+    }
+
+    async getToken(userId: number) {
```
Comment (WadasGrzegorz) [inline src/modules/redis/redis.service.ts:15]: just return this.cacheManager

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/modules/auth/auth.service.ts [hunk 6]
```diff
@@ -26,11 +29,30 @@ export class AuthService {
         return null
     }
 
-    login(user: User) {
-        const payload = { userName: user.userName, sub: user.userId }
+    async login(user: User) {
+        const uuid = uuidv4()
```
Comment (WadasGrzegorz) [inline src/modules/auth/auth.service.ts:33]: use userUUID instead create new one

File: src/modules/redis/redis.service.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common'
+import { Cache } from 'cache-manager'
+import { Jwt } from 'lib/enums'
+
+@Injectable()
+export class RedisService {
```
Comment (WadasGrzegorz) [inline src/modules/redis/redis.service.ts:6]: You should think about create this service more common

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/modules/auth/auth.service.ts (hunk 6 above)
Comment (paw9810) [reply src/modules/auth/auth.service.ts:33]: You mean to create userUUID in user entity?

File: src/modules/redis/redis.service.ts (hunk 7 above)
Comment (paw9810) [reply src/modules/redis/redis.service.ts:6]: So should I just create 3 basic methods and use them in other services?

Comment (WadasGrzegorz) [reply src/modules/redis/redis.service.ts:6]: Exactly, I recommend some methods like set, get, delete. If You do this you can use this module elsewhere

File: src/modules/auth/auth.service.ts (hunk 6 above)
Comment (WadasGrzegorz) [reply src/modules/auth/auth.service.ts:33]: Yes, please use for this ’@generated’ decorator from typeorm

**Comment by paw9810** [comment]
I have a problem with this approach with userUUID, because when I'm not generating a new one with every login, the validation will pass even with old tokens

File: src/modules/auth/auth.controller.ts [hunk 8]
```diff
@@ -33,6 +33,11 @@ export class AuthController {
         return this.authService.login(req.user)
     }
 
+    @Post('logout')
```
Comment (WadasGrzegorz) [inline src/modules/auth/auth.controller.ts:36]: use no-content

File: src/modules/redis/redis.service.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common'
+import { Cache } from 'cache-manager'
+
+@Injectable()
+export class RedisService {
+    constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}
+
+    async set(key: string, value: string, ttl: number) {
+        await this.cacheManager.set(key, value, { ttl })
```
Comment (WadasGrzegorz) [inline src/modules/redis/redis.service.ts:9]: await is not necessary just return

File: src/modules/redis/redis.service.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    async get(key: string) {
+        return this.cacheManager.get(key)
+    }
+
+    async del(key: string) {
+        await this.cacheManager.del(key)
```
Comment (WadasGrzegorz) [inline src/modules/redis/redis.service.ts:17]: aa

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/modules/auth/auth.controller.ts [hunk 11]
```diff
@@ -33,9 +34,10 @@ export class AuthController {
         return this.authService.login(req.user)
     }
 
+    @HttpCode(204)
```
Comment (WadasGrzegorz) [inline src/modules/auth/auth.controller.ts:37]: It's enum available in nest ```import { HttpStatus } from '@nestjs/common'``` and then ``` @HttpCode(HttpStatus.NO_CONTENT)```

File: src/modules/auth/strategies/jwt.strategy.ts [hunk 12]
```diff
@@ -20,10 +22,13 @@ export class JwtStrategy extends PassportStrategy(Strategy) {
     }
 
     async validate(payload: Payload) {
-        await this.authService.validateAccessToken(
-            payload.sub,
-            payload.tokenUUID
-        )
+        const user = await this.usersService.findOne(payload.userName)
+        if (user) {
```
Comment (WadasGrzegorz) [inline src/modules/auth/strategies/jwt.strategy.ts:26]: enter before if

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

**Review by WadasGrzegorz — APPROVED** [review]

**Pull request merged**